openldap/2.6/debian-11/rootfs/opt/bitnami/scripts/libopenldap.sh

```diff
@@ -0,0 +1,1045 @@
+        info "\t${ldif}"
+        debug_execute ldapadd -f "${ldif}" -H "${LDAP_LDAPI_URI}" -D "${LDAP_ADMIN_DN}" -w "${LDAP_ADMIN_PASSWORD}"
+    done
+}
+
+########################
+# OpenLDAP configure permissions
+# Globals:
+#   LDAP_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+ldap_configure_permissions() {
+  debug "Ensuring expected directories/files exist..."
+  for expected in "${ldap_expected_directories[@]}"; do
+      path="$(printenv $expected)"
+      ensure_dir_exists "$path" ${LDAP_DAEMON_USER} ${LDAP_DAEMON_GROUP}
+      if am_i_root; then
+          chmod -R g+rwX "$path"
+          chown -R "$LDAP_DAEMON_USER:$LDAP_DAEMON_GROUP" "$path"
+          info "Ensuring ${expected}=$path is $LDAP_DAEMON_USER:$LDAP_DAEMON_GROUP and 0775/drwxrwxr-x/g+rwX recursively"
+      else
+          groups=$(id -nG)
+          whoami="$(id -nu):${groups// /,}"
+          for item in $(find $path \! -type l); do
+              owgr="$(stat -c "%U:%G" "$item")"
+              if [[ "$owgr" != "$LDAP_DAEMON_USER:$LDAP_DAEMON_GROUP" ]]; then
+                  warn "${expected}=$item is $owgr rather than the expected $LDAP_DAEMON_USER:$LDAP_DAEMON_GROUP"
+              fi
+              perms="$(stat "$item" | grep -oP "(?<=Access: \()[^)]*")"
+              if [[ "$perms" =~ 07[^75]5/drwxrwxr-x ]]; then
+                  warn "${expected}=${item} (${owgr}) has permissions $perms process exec'ed by ${whoami}"
+              fi
+          done
+      fi
+  done
+}
+
+########################
+# Initialize OpenLDAP server
+# Globals:
+#   LDAP_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+ldap_initialize() {
+    info "Initializing OpenLDAP..."
+
+    ldap_configure_permissions
+    if are_dirs_empty "${LDAP_DATA_DIR}" "${LDAP_ONLINE_CONF_DIR}"; then
+        info "Setting up ${LDAP_VOLUME_DIR}/{data,slapd.d} config and data."
+        ensure_dir_exists "${LDAP_ONLINE_CONF_DIR}" ${LDAP_DAEMON_USER} ${LDAP_DAEMON_GROUP}
+        ensure_dir_exists "${LDAP_DATA_DIR}" ${LDAP_DAEMON_USER} ${LDAP_DAEMON_GROUP}
+
+        # Create OpenLDAP online configuration
+        ldap_create_online_configuration
+        ldap_start_bg
+        ldap_admin_credentials
+        info "Setting up optional config..."
+        if ! is_boolean_yes "$LDAP_ALLOW_ANON_BINDING"; then
+            ldap_disable_anon_binding
+        fi
+        if is_boolean_yes "$LDAP_ENABLE_TLS"; then
+            ldap_configure_tls
+        fi
+        # Initialize OpenLDAP with schemas/tree structure
+        if is_boolean_yes "$LDAP_ADD_SCHEMAS"; then
+            ldap_add_schemas
+        fi
+        if [[ -f "$LDAP_CUSTOM_SCHEMA_FILE" ]]; then
+            ldap_add_custom_schema
+        fi
+        if ! is_dir_empty "$LDAP_CUSTOM_SCHEMA_DIR"; then
+            ldap_add_custom_schemas
+        fi
+        # additional configuration
+        if [[ ! "$LDAP_PASSWORD_HASH" == "{SSHA}" ]]; then
```

Comment:
Move the `!` outside of the test for readability? Might also need to load an appropriate module to be able to do this.

Reply:
What module?  This could be another subtle difference between our build and Bitnami's.

Reply:
Password hash implementations might require a relevant module to be loaded. Which module depend on what the hash is. None of these tend to be compiled in statically.

Reply:
For argon2 I load that module now, for salt I add extra config.  Are there other special cases to encode?